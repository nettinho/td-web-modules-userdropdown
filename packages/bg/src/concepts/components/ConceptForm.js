import _ from "lodash/fp";
import React from "react";
import PropTypes from "prop-types";
import {
  Button,
  Container,
  Form,
  Header,
  Icon,
  Segment,
  Message
} from "semantic-ui-react";
import { bindActionCreators, compose } from "redux";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { injectIntl, FormattedMessage } from "react-intl";
import { RichTextEditor } from "@truedat/core/components";
import { required, ruleRunner, run } from "@truedat/core/validation";
import { DynamicForm, TemplateSelector } from "@truedat/df/components";
import {
  fetchDomainTemplates,
  clearDomainTemplates
} from "@truedat/df/routines";
import {
  conceptAction,
  fetchDomainConcepts,
  clearDomainConcepts
} from "../routines";
import DomainSelector from "./DomainSelector";
import ParentConceptSelector from "./ParentConceptSelector";

const actionKey = "create";

const staticFields = ["name", "description"];

const initialState = {
  name: "",
  description: {},
  content: {},
  domain: null,
  load_content: true,
  showStaticErrors: false,
  showDynamicErrors: false,
  template: {},
  validationStaticErrors: {},
  validationDynamicErrors: {}
};

const fieldValidations = formatMessage => {
  return [
    ruleRunner(
      "domain",
      formatMessage({ id: "domain.selector.label" }),
      required
    ),
    ruleRunner("name", formatMessage({ id: "concepts.props.name" }), required),
    ruleRunner(
      "description",
      formatMessage({ id: "concepts.props.description" }),
      required
    )
  ];
};

export class ConceptForm extends React.Component {
  static propTypes = {
    history: PropTypes.object,
    action: PropTypes.object,
    conceptAction: PropTypes.func,
    conceptActionLoading: PropTypes.string,
    fetchDomainConcepts: PropTypes.func,
    clearDomainConcepts: PropTypes.func,
    domains: PropTypes.array,
    parents: PropTypes.array,
    intl: PropTypes.object
  };

  state = initialState;

  componentWillUnmount() {
    const { clearDomainConcepts, clearDomainTemplates } = this.props;
    clearDomainTemplates();
    clearDomainConcepts();
  }

  handleDomainSelected = (e, { value }) => {
    const domain = value
      ? _.find(_.propEq("id", value))(this.props.domains)
      : {};
    const domain_id = domain.id;

    const { validationStaticErrors, showStaticErrors } = this.getValidations();

    this.setState({
      domain,
      domain_id,
      load_content: false,
      parent_id: null,
      validationStaticErrors,
      template: {},
      showStaticErrors
    });

    const { fetchDomainConcepts, fetchDomainTemplates } = this.props;
    fetchDomainTemplates({ domain_id });
    fetchDomainConcepts({ domain_id });
  };

  handleTemplateSelected = (e, data) => {
    const { value } = data;
    const template = value
      ? _.find(_.propEq("id", value))(this.props.templates)
      : {};
    const type = _.prop("name")(template);
    this.setState({ template, type });
  };

  handleOnParentConceptSearchChange = (e, data) => {
    const { searchQuery } = data;
    const { fetchDomainConcepts, clearDomainConcepts } = this.props;
    const { domain_id } = this.state;
    if (domain_id) {
      const query = _.isEmpty(searchQuery) ? undefined : searchQuery;
      fetchDomainConcepts({ domain_id, query });
    } else {
      clearDomainConcepts();
      this.setState({ parent_id: null });
    }
  };

  handleParentSelected = (e, { value }) => {
    this.setState({
      parent_id: value,
      load_content: true
    });
  };

  getValidations = () => {
    const {
      intl: { formatMessage }
    } = this.props;
    const validationStaticErrors = run(
      { ...this.state },
      fieldValidations(formatMessage)
    );
    const showStaticErrors = Object.keys(validationStaticErrors).length !== 0;
    return { validationStaticErrors, showStaticErrors };
  };

  handleEditorChange = ({ value }) => {
    this.handleChange(null, { name: "description", value });
  };

  handleChange = (e, { name, value }) => {
    e && e.preventDefault();

    if (staticFields.includes(name)) {
      const validationForm = { ...this.state };
      validationForm[name] = value;
      const {
        validationStaticErrors,
        showStaticErrors
      } = this.getValidations();
      this.setState({
        [name]: value,
        validationStaticErrors,
        showStaticErrors
      });
    }
  };

  updatedDfContent = (content, validations) => {
    const validationDynamicErrors = { ...validations };
    const showDynamicErrors = Object.keys(validationDynamicErrors).length !== 0;
    this.setState({
      content,
      validationDynamicErrors,
      showDynamicErrors
    });
  };

  handleSubmit = e => {
    e.preventDefault();
    const { action, conceptAction } = this.props;
    const business_concept_version = _.pick([
      "content",
      "domain_id",
      "parent_id",
      "type",
      ...staticFields
    ])(this.state);

    conceptAction({
      action: actionKey,
      ...action,
      business_concept_version
    });
  };

  listErrors = ({ validationStaticErrors, validationDynamicErrors }) =>
    Object.entries({
      ...validationStaticErrors,
      ...validationDynamicErrors
    }).map(x => x[1]);

  render() {
    const { formatMessage } = this.props.intl;
    const {
      history,
      domains,
      action,
      conceptActionLoading,
      parents
    } = this.props;

    const {
      name,
      description,
      showStaticErrors,
      showDynamicErrors,
      domain_id,
      template,
      content
    } = this.state;

    const loading =
      action && action.href && conceptActionLoading === action.href;

    return (
      <React.Fragment>
        <Container as={Segment} text>
          <Header as="h2">
            <Icon name="book" />
            <Header.Content>
              <FormattedMessage id="concepts.actions.create" />
            </Header.Content>
          </Header>
          <Form
            loading={loading}
            warning={showStaticErrors || showDynamicErrors}
          >
            <DomainSelector
              onChange={this.handleDomainSelected}
              domains={domains}
            />
            <TemplateSelector
              selectedValue={template.id}
              onChange={this.handleTemplateSelected}
            />
            <ParentConceptSelector
              onChange={this.handleParentSelected}
              onSearchChange={this.handleOnParentConceptSearchChange}
            />
            <Form.Input
              label={formatMessage({ id: "concepts.props.name" })}
              name="name"
              value={name}
              onChange={this.handleChange}
              required
            />
            <RichTextEditor
              label={formatMessage({ id: "concepts.props.description" })}
              name="description"
              value={description}
              onChange={this.handleEditorChange}
              required
            />

            <DynamicForm
              template={template}
              updatedContent={this.updatedDfContent}
              content={content}
            />

            <Message
              warning
              header={formatMessage({ id: "validation.header.message.error" })}
              list={this.listErrors(this.state)}
            />
            <Button
              primary
              onClick={this.handleSubmit}
              content={<FormattedMessage id="actions.create" />}
              disabled={!action}
            />
            <Button
              default
              onClick={() => history.goBack()}
              content={<FormattedMessage id="actions.cancel" />}
            />
          </Form>
        </Container>
      </React.Fragment>
    );
  }
}

const mapStateToProps = state => ({
  domains: state.domains,
  templates: state.domainTemplates,
  conceptActionLoading: state.conceptActionLoading,
  action: _.get(actionKey)(state.conceptsActions)
});

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators(
    {
      conceptAction,
      fetchDomainTemplates,
      clearDomainTemplates,
      fetchDomainConcepts,
      clearDomainConcepts
    },
    dispatch
  )
});

export default compose(
  withRouter,
  injectIntl,
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(ConceptForm);
