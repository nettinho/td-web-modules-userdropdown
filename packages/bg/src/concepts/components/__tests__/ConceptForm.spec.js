import React from "react";
import { shallowWithIntl } from "intl-stub";
import { ConceptForm } from "../ConceptForm";

describe("<ConceptForm />", () => {
  const conceptAction = jest.fn();
  const fetchTemplates = jest.fn();
  const domains = [{ id: 1, name: "Domain1" }, { id: 2, name: "Domain 2" }];
  const conceptActionLoading = "some action";
  const action = {};
  const template = {
    name: "Template1",
    id: 5,
    content: [
      {
        type: "string",
        required: false,
        name: "field1",
        label: "S1 String"
      },
      {
        values: ["Option 1", "Option 2", "Option 3"],
        type: "list",
        required: true,
        name: "field7",
        label: "List",
        group: "Group 1",
        default: "Option 2"
      }
    ]
  };

  const props = {
    conceptAction,
    fetchTemplates,
    domains,
    conceptActionLoading,
    action,
    template
  };

  it("matches the latest snapshot", () => {
    const wrapper = shallowWithIntl(<ConceptForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
