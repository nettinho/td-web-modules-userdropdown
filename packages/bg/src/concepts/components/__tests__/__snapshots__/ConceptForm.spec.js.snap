// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConceptForm /> matches the latest snapshot 1`] = `
<React.Fragment>
  <Container
    as={[Function]}
    text={true}
  >
    <Header
      as="h2"
    >
      <Icon
        as="i"
        name="book"
      />
      <HeaderContent>
        <FormattedMessage
          id="concepts.actions.create"
          values={Object {}}
        />
      </HeaderContent>
    </Header>
    <Form
      as="form"
      warning={false}
    >
      <InjectIntl(Connect(DomainSelector))
        domains={
          Array [
            Object {
              "id": 1,
              "name": "Domain1",
            },
            Object {
              "id": 2,
              "name": "Domain 2",
            },
          ]
        }
        onChange={[Function]}
      />
      <InjectIntl(Connect(TemplateSelector))
        onChange={[Function]}
      />
      <InjectIntl(Connect(ParentConceptSelector))
        onChange={[Function]}
        onSearchChange={[Function]}
      />
      <FormInput
        as={[Function]}
        control={[Function]}
        label="concepts.props.name"
        name="name"
        onChange={[Function]}
        required={true}
        value=""
      />
      <RichTextEditor
        label="concepts.props.description"
        name="description"
        onChange={[Function]}
        required={true}
        value={Object {}}
      />
      <DynamicForm
        content={Object {}}
        template={Object {}}
        updatedContent={[Function]}
      />
      <Message
        header="validation.header.message.error"
        list={Array []}
        warning={true}
      />
      <Button
        as="button"
        content={
          <FormattedMessage
            id="actions.create"
            values={Object {}}
          />
        }
        disabled={false}
        onClick={[Function]}
        primary={true}
        role="button"
      />
      <Button
        as="button"
        content={
          <FormattedMessage
            id="actions.cancel"
            values={Object {}}
          />
        }
        default={true}
        onClick={[Function]}
        role="button"
      />
    </Form>
  </Container>
</React.Fragment>
`;
